File: leetcode/152-maximum-product-subarray/main.java
import java.lang.Math;

class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        int min = nums[0];
        int max = min;
        int ret = min;

        for (int i = 1; i < nums.length; i++) {
            if (nums[i] < 0) {
                int v = max;
                max = min;
                min = v;
            }

            max = Math.max(nums[i], nums[i] * max);
            min = Math.min(nums[i], nums[i] * min);
            ret = Math.max(max, Math.max(min, ret));

        }
        return ret;
    }
}

File: leetcode/152-maximum-product-subarray/main.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
      if (nums.size() == 0) {
        return 0;
      }

      int mn = nums[0];
      int mx = mn;
      int ret = mn;

      for (int i = 1; i < nums.size(); i++) {
        if (nums[i] < 0) {
          int v = mx;
          mx = mn;    
          mn = v;
        }

        mx = max(nums[i], nums[i] * mx);
        mn = min(nums[i], nums[i] * mn);
        ret = max(mx, max(ret, mn));
      }

      return ret;
        
    }
};
